hypermath/src/pga/blade.rs

```diff
@@ -311,6 +317,74 @@ impl Blade {
         }
     }
 
+    /// Returns the projection of the product of the blades to the specified grade.
+    pub fn product_grade(lhs: &Self, rhs: &Self, grade: u8) -> Self {
+        let ndim = std::cmp::max(lhs.ndim, rhs.ndim);
+
+        let mut ret = Self::zero(ndim, grade);
+        for l in lhs.terms() {
+            for r in rhs.terms() {
+                let Some(term) = Term::geometric_product(l, r) else {
+                    continue;
+                };
+                if term.grade() == grade {
+                    ret += term;
+                }
+            }
+        }
+        ret
+    }
+
+    /// Returns the projection of the product of the blades to the grade 0.
+    /// May differ from dot by a sign.
+    pub fn product_scalar(lhs: &Self, rhs: &Self) -> Float {
+        let mut ret = 0.0;
+        for l in lhs.terms() {
+            for r in rhs.terms() {
+                let Some(term) = Term::geometric_product(l, r) else {
+                    continue;
+                };
+                if term.grade() == 0 {
+                    ret += term.coef;
+                }
+            }
+        }
+        ret
+    }
+    /// Returns the projection of the product of the blades to the specified grade.
+    pub fn multi_product_grade(blades: impl IntoIterator<Item = Self>, grade: u8) -> Self {
+        let blades = blades.into_iter().collect_vec();
+        let ndim = blades.iter().map(|b| b.ndim).max().unwrap_or(0);
+
+        let mut ret = Self::zero(ndim, grade);
+
+        'a: for terms in blades
+            .into_iter()
+            .map(|b| b.terms().collect_vec())
+            .multi_cartesian_product()
+        {
+            let mut term = Term::scalar(1.0);
+            for t in terms {
```

Comment:
surely this inner loop can be expressed using a functional reduce? maybe it's less nice that way